### DISCUSSION TITLE
Add `router.key` to identify history

### DISCUSSION THREAD

**Discussion by AkifumiSato** [vercel/next.js#47242] [Ideas]
### Goals

1. History Identification
1. Restoration of UI state

### Background

The state restoration based on history is a very important UX, and this requires that the Next.js history be uniquely identifiable.

- @rauchg also previously mentioned the importance of browser backs.
  - https://rauchg.com/2014/7-principles-of-rich-web-applications#dont-break-history-enhance-it
  - > Don't break history, enhance it
- The [Navigation API](https://github.com/WICG/navigation-api#the-current-entry) also exposes keys in the same way.

### Proposal

Add `router.key` to both `app` and `page`.

**Comment by AkifumiSato** [comment]
As was the case with `pages`, the `app router` is still pushing/replacing `history.state`, so restoring state based on history is not feasible unless Next.js exposes the key.
https://github.com/vercel/next.js/blob/285e77541f/packages/next/src/client/components/app-router.tsx#L113-L115

**Comment by sumiren** [comment]
I also need key in `history.state` as pages dir.

**Comment by sumiren** [comment]
We cannot migrate to App Router because App Router has no key in `history.state`.

We have custom routing logic which manage history per tab.
In order to restore the history when browser-back, we need `router.key`  to save with our history to client storage.

**Comment by kyeshmz** [comment]
Is this something being handled alongside fetch api? @leerob

**Comment by koenoe** [comment]
Not sure if still helpful for anyone, but adding this to the root layout works great for me!
```
<script
  id="history-state-key"
  suppressHydrationWarning
  // eslint-disable-next-line react/no-danger
  dangerouslySetInnerHTML={{
    __html: `
      var orgPushState = window.history.pushState;
      var orgReplaceState = window.history.replaceState;

      function ensureStateKey(state) {
          if (!state || !state.key) {
              var key = Math.random().toString(32).slice(2);
              state = state || {};
              state.key = key;
          }
          return state;
      }

      window.history.pushState = function (state, unused, url) {
          state = ensureStateKey(state);
          orgPushState.call(window.history, state, unused, url);
      };

      window.history.replaceState = function (state, unused, url) {
          state = ensureStateKey(state);
          orgReplaceState.call(window.history, state, unused, url);
      };
    `,
  }}
/>
        ```
